### PR TITLE
Update CVE feed layouts for new JSON feed format

### PR DESCRIPTION
**Pull request by mtardy**
This PR fixes the layout for the CVE feed for the new JSON format, **to be merged after https://github.com/kubernetes/sig-security/pull/75 and https://github.com/kubernetes/sig-security/pull/76 are merged**. While these are not merged and deployed, the build will fail because the JSON format is different.

- https://github.com/kubernetes/sig-security/pull/75 changes some of the fields' names to comply with JSON feed specification.
- https://github.com/kubernetes/sig-security/pull/76 adds the `_kubernetes_io.updated_at` root field used to display the last update date in the caption of the CVE table.

**Note that PR 75 was merged into PR 76 that bundles all the changes for simplicity.**

Note that `{{ getJSON .Site.Params.cveFeedBucket | jsonify }}` can look useless but it's to make the JSON file looks "unpretty" like previously (with no space and no indentation). If we don't care about this we can just replace it with `{{ os.ReadFile .Site.Params.cveFeedBucket }}`. There was some note about this Hugo issue https://github.com/gohugoio/hugo/issues/7229 but I did not encounter this, you can check using `curl` and `grep <` to see.

### Testing this PR locally

1. Generate the new JSON CVE feed file with the new bash and python scripts from PR 75 and 76, or just download this version generated (with only the four latest CVEs in order to reduce the size of the file) and put the file under `static/official-cve-feed.json`.
    <details>
      <summary><code>official-cve-feed.json</code> extract new version according to PR 75 and 76 as of 2022-12-20T15:12:31Z</summary>

    ```json
    {
    "version": "https://jsonfeed.org/version/1.1",
    "title": "Auto-refreshing Official CVE Feed",
    "home_page_url": "https://kubernetes.io",
    "feed_url": "https://kubernetes.io/docs/reference/issues-security/official-cve-feed/index.json",
    "description": "Auto-refreshing official CVE feed for Kubernetes repository",
    "authors": [
        {
            "name": "Kubernetes Community",
            "url": "https://www.kubernetes.dev"
        }
    ],
    "_kubernetes_io": {
        "updated_at": "2022-12-20T15:12:31Z"
    },
    "items": [
        {
            "content_text": "CVSS Rating: [CVSS:3.1/AV:N/AC:H/PR:H/UI:N/S:U/C:H/I:H/A:H](https://www.first.org/cvss/calculator/3.1#CVSS:3.1/AV:N/AC:H/PR:H/UI:N/S:U/C:H/I:H/A:H)\r\n\r\nA security issue was discovered in Kubernetes where users may have access to secure endpoints in the control plane network. Kubernetes clusters are only affected if an untrusted user can modify Node objects and send proxy requests to them.\r\n\r\nKubernetes supports node proxying, which allows clients of kube-apiserver to access endpoints of a Kubelet to establish connections to Pods, retrieve container logs, and more. While Kubernetes already validates the proxying address for Nodes, a bug in kube-apiserver made it possible to bypass this validation. Bypassing this validation could allow authenticated requests destined for Nodes to to the API server's private network.\r\n\r\n### Am I vulnerable?\r\n\r\nClusters are affected by this vulnerability if there are endpoints that the kube-apiserver has connectivity to that users should not be able to access. This includes:\r\n\r\n- kube-apiserver is in a separate network from worker nodes\r\n- localhost services\r\n\r\nmTLS services that accept the same client certificate as nodes may be affected. The severity of this issue depends on the privileges & sensitivity of the exploitable endpoints.\r\n\r\nClusters that configure the egress selector to use a proxy for cluster traffic may not be affected.\r\n\r\n\r\n#### Affected Versions\r\n\r\n- Kubernetes kube-apiserver <= v1.25.3\r\n- Kubernetes kube-apiserver <= v1.24.7\r\n- Kubernetes kube-apiserver <= v1.23.13\r\n- Kubernetes kube-apiserver <= v1.22.15\r\n\r\n### How do I mitigate this vulnerability?\r\n\r\nUpgrading the **kube-apiserver** to a fixed version mitigates this vulnerability.\r\n\r\nAside from upgrading, configuring an [egress proxy for egress to the cluster network](https://kubernetes.io/docs/tasks/extend-kubernetes/setup-konnectivity/) can mitigate this vulnerability.\r\n\r\n#### Fixed Versions\r\n\r\n- Kubernetes kube-apiserver v1.25.4\r\n- Kubernetes kube-apiserver v1.24.8\r\n- Kubernetes kube-apiserver v1.23.14\r\n- Kubernetes kube-apiserver v1.22.16\r\n\r\n**Fix impact:** In some cases, the fix can break clients that depend on the nodes/proxy subresource, specifically if a kubelet advertises a localhost or link-local address to the Kubernetes control plane.\r\n\r\n### Detection\r\n\r\nNode create & update requests may be included in the Kubernetes audit log, and can be used to identify requests for IP addresses that should not be permitted. Node proxy requests may also be included in audit logs.\r\n\r\nIf you find evidence that this vulnerability has been exploited, please contact security@kubernetes.io\r\n\r\n#### Acknowledgements\r\n\r\nThis vulnerability was reported by Yuval Avrahami of Palo Alto Networks.\r\n\r\n<!-- labels -->\r\n/area security\r\n/kind bug\r\n/committee security-response\r\n/label official-cve-feed\r\n/sig api-machinery\r\n/area apiserver",
            "date_published": "2022-11-08T21:33:26Z",
            "external_url": "https://www.cve.org/cverecord?id=CVE-2022-3294",
            "id": "CVE-2022-3294",
            "summary": "Node address isn't always verified when proxying",
            "url": "https://github.com/kubernetes/kubernetes/issues/113757",
            "_kubernetes_io": {
                "google_group_url": "https://groups.google.com/g/kubernetes-announce/search?q=CVE-2022-3294",
                "issue_number": 113757
            }
        },
        {
            "content_text": "CVSS Rating: [CVSS:3.0/AV:N/AC:L/PR:L/UI:N/S:U/C:H/I:N/A:N](https://www.first.org/cvss/calculator/3.0#CVSS:3.0/AV:N/AC:L/PR:L/UI:N/S:U/C:H/I:N/A:N)\r\n\r\nA security issue was discovered in Kubernetes where users authorized to list or watch one type of namespaced custom resource cluster-wide can read custom resources of a different type in the same API group without authorization.\r\n\r\n### Am I vulnerable?\r\n\r\nClusters are impacted by this vulnerability if all of the following are true:\r\n- There are 2+ CustomResourceDefinitions sharing the same API group\r\n- Users have cluster-wide list or watch authorization on one of those custom resources.\r\n- The same users are not authorized to read another custom resource in the same API group.\r\n\r\n#### Affected Versions\r\n\r\n- Kubernetes kube-apiserver <= v1.25.3\r\n- Kubernetes kube-apiserver <= v1.24.7\r\n- Kubernetes kube-apiserver <= v1.23.13\r\n- Kubernetes kube-apiserver <= v1.22.15\r\n\r\n### How do I mitigate this vulnerability?\r\n\r\nUpgrading the kube-apiserver to a fixed version mitigates this vulnerability.\r\n\r\nPrior to upgrading, this vulnerability can be mitigated by avoiding granting cluster-wide list and watch permissions.\r\n\r\n#### Fixed Versions\r\n\r\n- Kubernetes kube-apiserver v1.25.4\r\n- Kubernetes kube-apiserver v1.24.8\r\n- Kubernetes kube-apiserver v1.23.14\r\n- Kubernetes kube-apiserver v1.22.16\r\n\r\n### Detection\r\n\r\nRequests containing `..` in the request path are a likely indicator of exploitation. Request paths may be captured in API audit logs, or in kube-apiserver HTTP logs.\r\n\r\nIf you find evidence that this vulnerability has been exploited, please contact security@kubernetes.io\r\n\r\n#### Acknowledgements\r\n\r\nThis vulnerability was reported by Richard Turnbull of NCC Group as part of the Kubernetes Audit.\r\n\r\n<!-- labels -->\r\n/area security\r\n/kind bug\r\n/committee security-response\r\n/label official-cve-feed\r\n/sig api-machinery\r\n/area apiserver\r\n",
            "date_published": "2022-11-08T21:33:07Z",
            "external_url": "https://www.cve.org/cverecord?id=CVE-2022-3162",
            "id": "CVE-2022-3162",
            "summary": "Unauthorized read of Custom Resources",
            "url": "https://github.com/kubernetes/kubernetes/issues/113756",
            "_kubernetes_io": {
                "google_group_url": "https://groups.google.com/g/kubernetes-announce/search?q=CVE-2022-3162",
                "issue_number": 113756
            }
        },
        {
            "content_text": "CVSS Rating: [CVSS:3.1/AV:N/AC:H/PR:H/UI:R/S:C/C:L/I:L/A:L](https://www.first.org/cvss/calculator/3.1#CVSS:3.1/AV:N/AC:H/PR:H/UI:R/S:C/C:L/I:L/A:L) (5.1, medium)\r\n\r\nA security issue was discovered in kube-apiserver that allows an aggregated API server to redirect client traffic to any URL.  This could lead to the client performing unexpected actions as well as forwarding the client's API server credentials to third parties.\r\n\r\nThis issue has been rated medium and assigned CVE-2022-3172\r\n\r\n### Am I vulnerable?\r\n\r\nAll Kubernetes clusters with the following versions that are running aggregated API servers are impacted.  To identify if you have aggregated API servers configured, run the following command:\r\n\r\n```shell\r\nkubectl get apiservices.apiregistration.k8s.io -o=jsonpath='{range .items[?(@.spec.service)]}{.metadata.name}{\"\\n\"}{end}'\r\n```\r\n\r\n#### Affected Versions\r\n\r\n- kube-apiserver v1.25.0\r\n- kube-apiserver v1.24.0 - v1.24.4\r\n- kube-apiserver v1.23.0 - v1.23.10\r\n- kube-apiserver v1.22.0 - v1.22.13\r\n- kube-apiserver <= v1.21.14\r\n\r\n### How do I mitigate this vulnerability?\r\n\r\nAside from upgrading, no direct mitigation is available.\r\n\r\nAggregated API servers are a trusted part of the Kubernetes control plane, and configuring them is a privileged administrative operation.  Ensure that only trusted cluster administrators are allowed to create or modify `APIService` configuration, and follow security best practices with any aggregated API servers that may be in use.\r\n\r\n#### Fixed Versions\r\n\r\n- kube-apiserver v1.25.1 - fixed by #112330\r\n- kube-apiserver v1.24.5 - fixed by #112331\r\n- kube-apiserver v1.23.11 - fixed by #112358\r\n- kube-apiserver v1.22.14 - fixed by #112359\r\n\r\n**Fix impact:** The fix blocks all 3XX responses from aggregated API servers by default.  This may disrupt an aggregated API server that relies on redirects as part of its normal function.  If all current and future aggregated API servers are considered trustworthy and redirect functionality is required, set the `--aggregator-reject-forwarding-redirect` Kubernetes API server flag to `false` to restore the previous behavior.\r\n\r\nTo upgrade, refer to the documentation: https://kubernetes.io/docs/tasks/administer-cluster/cluster-upgrade\r\n\r\n### Detection\r\n\r\nKubernetes audit log events indicate the HTTP status code sent to the client via the `responseStatus.code` field.  This can be used to detect if an aggregated API server is redirecting clients.\r\n\r\nIf you find evidence that this vulnerability has been exploited, please contact security@kubernetes.io\r\n\r\n#### Acknowledgements\r\n\r\nThis vulnerability was reported by Nicolas Joly & Weinong Wang @weinong from Microsoft.\r\n\r\nThe issue was fixed and coordinated by Di Jin @jindijamie @enj @liggitt @lavalamp @deads2k and @puerco.\r\n\r\n/area security\r\n/kind bug\r\n/committee security-response\r\n/label official-cve-feed\r\n/sig api-machinery\r\n/area apiserver\r\n/triage accepted",
            "date_published": "2022-09-16T13:14:50Z",
            "external_url": "https://www.cve.org/cverecord?id=CVE-2022-3172",
            "id": "CVE-2022-3172",
            "summary": "Aggregated API server can cause clients to be redirected (SSRF)",
            "url": "https://github.com/kubernetes/kubernetes/issues/112513",
            "_kubernetes_io": {
                "google_group_url": "https://groups.google.com/g/kubernetes-announce/search?q=CVE-2022-3172",
                "issue_number": 112513
            }
        },
        {
            "content_text": "A security issue was discovered in Kubernetes that could allow  Windows workloads to run as `ContainerAdministrator` even when those workloads set the `runAsNonRoot` option to `true`.\r\n\r\nThis issue has been rated low and assigned CVE-2021-25749\r\n\r\n### Am I vulnerable?\r\n\r\nAll Kubernetes clusters with following versions, running Windows workloads with `runAsNonRoot` are impacted\r\n\r\n#### Affected Versions\r\n\r\n- kubelet v1.20 - v1.21\r\n- kubelet v1.22.0 - v1.22.13\r\n- kubelet v1.23.0 - v1.23.10\r\n- kubelet v1.24.0 - v1.24.4\r\n\r\n### How do I mitigate this vulnerability?\r\n\r\nThere are no known mitigations to this vulnerability.\r\n\r\n#### Fixed Versions\r\n\r\n- kubelet v1.22.14\r\n- kubelet v1.23.11\r\n- kubelet v1.24.5\r\n- kubelet v1.25.0\r\n\r\n\r\nTo upgrade, refer to this documentation _For core Kubernetes:_ https://kubernetes.io/docs/tasks/administer-cluster/cluster-upgrade/\r\n\r\n### Detection\r\n\r\nKubernetes Audit logs may indicate if the user name was misspelled to bypass the restriction placed on which user is a pod allowed to run as.\r\n\r\nIf you find evidence that this vulnerability has been exploited, please contact security@kubernetes.io\r\n\r\n#### Additional Details\r\n\r\nSee the GitHub issue for more details: https://github.com/kubernetes/kubernetes/issues/112192 \r\n\r\n#### Acknowledgements\r\n\r\nThis vulnerability was reported and fixed by Mark Rosetti (@marosset)\r\n",
            "date_published": "2022-09-01T21:02:01Z",
            "external_url": "https://www.cve.org/cverecord?id=CVE-2021-25749",
            "id": "CVE-2021-25749",
            "summary": "`runAsNonRoot` logic bypass for Windows containers",
            "url": "https://github.com/kubernetes/kubernetes/issues/112192",
            "_kubernetes_io": {
                "google_group_url": "https://groups.google.com/g/kubernetes-announce/search?q=CVE-2021-25749",
                "issue_number": 112192
            }
        }
      ]
    }

    ```
    
  </details>

2. Modify `config.toml` to use the local file.

    ```diff
    diff --git a/config.toml b/config.toml
    index c34cfa20d2..105d486b06 100644
    --- a/config.toml
    +++ b/config.toml
    @@ -169,7 +169,7 @@ algolia_docsearch = false
     offlineSearch = false
    
     # Official CVE feed bucket URL
    -cveFeedBucket = "https://storage.googleapis.com/k8s-cve-feed/official-cve-feed.json"
    +cveFeedBucket = "/static/official-cve-feed.json"
    
     [params.pushAssets]
     css = [
    ```

#### Preview

The HTML page looks very similar with just a new caption indicating the last update date.

image

### CODE REVIEW

File: layouts/shortcodes/cve-feed.html [hunk 1]
```diff
@@ -1,5 +1,6 @@
 <table class="security-cves">
-  <caption>{{ T "cve_table" }}</caption>
+  {{ $feed := getJSON .Site.Params.cveFeedBucket }}
+  <caption>{{ T "cve_table" }} (updated on {{ $feed._kubernetes_io.updated_at | time.Format "02 Jan 2006 at 15:04:05 MST" }})</caption>
```
Comment (sftim) [inline layouts/shortcodes/cve-feed.html:3]: Please make string “updated on” localizable.

Comment (sftim) [inline layouts/shortcodes/cve-feed.html:3]: Use a localized / localizable time format, not `"02 Jan 2006 at 15:04:05 MST"` (which embeds an English word)

**Review by sftim — COMMENTED** [review]
Optionally, check that `"version": "https://jsonfeed.org/version/1.1"` in the fetched feed, and fail the build if this doesn't match.

File: layouts/shortcodes/cve-feed.html (hunk 1 above)
Comment (mtardy) [reply layouts/shortcodes/cve-feed.html:3]: Yes! And I should clean up the no longer used localization strings.

Comment (PushkarJ) [reply layouts/shortcodes/cve-feed.html:3]: Can we replace "updated on" with "last updated: " ?

Comment (PushkarJ) [reply layouts/shortcodes/cve-feed.html:3]: @mtardy can you try the options from here to resolve @sftim's suggestion: https://gohugo.io/functions/dateformat/ ?

Comment (PushkarJ) [reply layouts/shortcodes/cve-feed.html:3]: @mtardy another example: https://github.com/kubernetes/website/blob/3ca34b9a3be454055ba234f1d2ff7d55809b5040/layouts/index.rss.xml#L24

Comment (sftim) [reply layouts/shortcodes/cve-feed.html:3]: Try:
```html
  <caption>{{ T "cve_table" }} (updated on {{ $feed._kubernetes_io.updated_at | time.Format ( T "cve_table_date_format" ) }})</caption>
```

Comment (mtardy) [reply layouts/shortcodes/cve-feed.html]: Done here https://github.com/kubernetes/website/pull/38579/commits/7f13b94edfc3302f351df03967bb2ab4edf3cfbc#diff-2d1b915ccd9669900a511913db7cc235638bbd7916ba066742477288fd0c503cR3.

Comment (mtardy) [reply layouts/shortcodes/cve-feed.html]: Yes, that's perfect thanks for your help!

**Comment by mtardy** [comment]
We thought with @PushkarJ that it would be better to put the caption at the top of the table in order for it to be more noticeable. I took the liberty to add a property of CSS directly in the shortcode HTML template (since the class specified seems not to exist).

See here https://github.com/kubernetes/website/pull/38579/files#diff-2d1b915ccd9669900a511913db7cc235638bbd7916ba066742477288fd0c503cR3.

**Comment by mtardy** [comment]
> Optionally, check that `"version": "https://jsonfeed.org/version/1.1"` in the fetched feed, and fail the build if this doesn't match.

Yes! Why not, I added this stuff https://github.com/kubernetes/website/pull/38579/commits/32442ffcc556d77b7034f89839779b20a583118b.

**Comment by sftim** [comment]
The build error is:
```
Error building site: "/opt/build/repo/content/en/docs/reference/issues-security/official-cve-feed.md:31:1": failed to render shortcode "cve-feed": failed to process shortcode: "/opt/build/repo/layouts/shortcodes/cve-feed.html:2:14": execute of template failed: template: shortcodes/cve-feed.html:2:14: executing "shortcodes/cve-feed.html" at <$feed.version>: can't evaluate field version in type []interface {}
```

**Comment by sftim** [comment]
Make sure that the upstream feed includes that change!

**Comment by PushkarJ** [comment]
/label tide/merge-method-squash
/lgtm

@sftim please `/approve` and keep this on `/hold` so when @mtardy is ready to merge both he can merge them together.

File: data/i18n/en/en.toml [hunk 2]
```diff
@@ -70,6 +58,15 @@ other = "Official Kubernetes CVE List"
...
 
+[cve_table_date_before]
+other = "(last updated: "
+
+[cve_table_date_format]
+other = "02 Jan 2006 15:04:05 MST"
+
+[cve_table_date_after]
+other = ")"
+
```
Comment (sftim) [inline data/i18n/en/en.toml:69]: The collation order is wrong here. In English, `t` collates before `u`.

**Comment by sftim** [comment]
/remove-label tide/merge-method-squash

It's useful to note this history. If there were 22 commits I'd think differently, mind.

File: data/i18n/en/en.toml (hunk 2 above)
Comment (mtardy) [reply data/i18n/en/en.toml:69]: I'm sorry I really tried but I don't understand what you mean 😞, is there a problem with "last updated"?

Comment (sftim) [reply data/i18n/en/en.toml:69]: Try this order for the keys:

* `cve_id`
* `cve_issue_url`
* `cve_json_external_url`
* `cve_json_id`
* `cve_json_summary`
* `cve_json_url`
* `cve_summary`
* `cve_table`
* ~`cve_url`~
* `cve_table_date_before`
* `cve_table_date_format`
* `cve_table_date_after`
* **`cve_url`**

Comment (mtardy) [reply data/i18n/en/en.toml:69]: aaaaah I did not even notice they were ordered alphabetically ok ok

Comment (mtardy) [reply data/i18n/en/en.toml:69]: Ok I just removed `cve_url` because it was unused anyway

**Comment by kbhawkey** [comment]
@mtardy . Are you waiting for another pull request to merge?

**Comment by mtardy** [comment]
> @mtardy . Are you waiting for another pull request to merge?

@kbhawkey Not really, I would need someone to approve this one, even though it's currently not building, here's why:
This PR is mutually dependent on this https://github.com/kubernetes/sig-security/pull/76. But this one has a lgtm and approve on hold.
I would need an lgtm and an approve here on hold here as well so that I can manage to unhold and merge them approximately at the same time so that the amount of downtime is minimal.

**Comment by PushkarJ** [comment]
+1 to what @mtardy said

File: data/i18n/en/en.toml [hunk 3]
```diff
@@ -49,26 +49,20 @@ other = "CVE ID"
...
 [cve_summary]
 other = "Issue Summary"
 
 [cve_table]
 other = "Official Kubernetes CVE List"
 
-[cve_url]
-other = "CVE URL"
+[cve_table_date_before]
+other = "(last updated: "
```
Comment (kbhawkey) [inline data/i18n/en/en.toml:59]: This is a bit unusual
I'll look at how you are constructing the string

Comment (mtardy) [reply data/i18n/en/en.toml:59]: I was thinking on the contrary that it was how we usually do these things, but we can do whatever you think is better.
https://github.com/kubernetes/website/blob/206231dcac5fbcce000bbc21a01a6a9806b836d4/data/i18n/en/en.toml#L275-L279

File: layouts/shortcodes/cve-feed.html [hunk 4]
```diff
@@ -1,19 +1,23 @@
+{{ $feed := getJSON .Site.Params.cveFeedBucket }}
+{{ if ne $feed.version "https://jsonfeed.org/version/1.1" }}
+  {{ errorf "Build Failed. CVE feed does not comply with JSON feed v1.1" }}
```
[the remark this answers is not in the document]
Comment (mtardy) [reply layouts/shortcodes/cve-feed.html:3]: Ok makes sense. This is a bit in contradiction with this previous review https://github.com/kubernetes/website/pull/38579#pullrequestreview-1224764138, @sftim do you have an opinion on that as well?

File: data/i18n/en/en.toml (hunk 3 above)
Comment (sftim) [reply data/i18n/en/en.toml:59]: I prefer to have the localized string be a _format string_:
```toml
[cve_table_date_with_label]
other = "(last updated: %s)"
```
and then use a string formatter to include the localized date. That's equivalent, but the placeholder makes it easier for localization teams to understand.

**Comment by kbhawkey** [comment]
Hi, @mtardy .
I'm not completely clear on how these dependencies are set up, but if possible eventually it would be a good idea to decouple the two repos (let the website build with a copy of the spec that would enable the website to always build).

Anyway, the  currently published page is: `https://kubernetes.io/docs/reference/issues-security/official-cve-feed/`, not sure what the new page will look like.

/approve

File: data/i18n/en/en.toml (hunk 3 above)
Comment (mtardy) [reply data/i18n/en/en.toml:59]: I'll update this in the follow-up PR https://github.com/kubernetes/website/pull/39513 :)

File: layouts/shortcodes/cve-feed.html (hunk 4 above)
[the remark this answers is not in the document]
Comment (mtardy) [reply layouts/shortcodes/cve-feed.html:3]: Yep we'll look at this here https://github.com/kubernetes/website/pull/39513 next!

**Comment by mtardy** [comment]
> Hi, @mtardy . I'm not completely clear on how these dependencies are set up, but if possible eventually it would be a good idea to decouple the two repos (let the website build with a copy of the spec that would enable the website to always build).
> 
> Anyway, the currently published page is: `https://kubernetes.io/docs/reference/issues-security/official-cve-feed/`, not sure what the new page will look like.
> 
> /approve

The workflow is the following: a script in kubernetes/sig-security is picked up by a job in k8s infra, [scheduled to run every 2 hours](https://testgrid.k8s.io/sig-security-cve-feed#auto-refreshing-official-cve-feed), and the website retrieves the file created by this job to update its internal JSON data, required to build this page.

I just merged the script side, the next job should run at 11:09 CET and then the preview of this PR should work, and the old live page will fail if I do not merge this new one once it picks up the new JSON :)

**Comment by mtardy** [comment]
I tested locally and the web page works with the new output format, let's merge.
/unhold

**Comment by mtardy** [comment]
Does this work on website?
/retest

**Pull request merged**